File: itmo-2s-golang-hw3/pkg/range/empty_range.go
package rangeI

type EmptyRange struct {
}

func (r EmptyRange) Length() int {
	return 0
}

func (r EmptyRange) Intersect(other RangeInt) RangeInt {
	return EmptyRange{}
}

func (r EmptyRange) Union(other RangeInt) (RangeInt, bool) {
	return other, true
}

func (r EmptyRange) IsEmpty() bool {
	return true
}

func (r EmptyRange) ContainsInt(i int) bool {
	return false
}

func (r EmptyRange) ContainsRange(other RangeInt) bool {
	return other.IsEmpty()
}

func (r EmptyRange) IsIntersect(other RangeInt) bool {
	return false
}

func (r EmptyRange) ToSlice() []int {
	return []int{}
}

func (r EmptyRange) Minimum() (int, bool) {
	return 0, false
}

func (r EmptyRange) Maximum() (int, bool) {
	return 0, false
}

func (r EmptyRange) ToString() string {
	return ""
}

File: itmo-2s-scala-hw3/src/main/scala/range/EmptyRange.scala
package range

object EmptyRange extends Range {
  override def length: Int = 0

  override def intersect(other: Range): Range = EmptyRange

  override def union(other: Range): Option[Range] =
    if (other.isEmpty) Some(Range.empty)
    else Some(Range.fromInterval(other.minimum.get, other.maximum.get))

  override def isEmpty: Boolean = true

  override def contains(i: Int): Boolean = false

  override def contains(other: Range): Boolean = other.isEmpty

  override def isIntersect(other: Range): Boolean = false

  override def toList: List[Int] = List.empty

  override def minimum: Option[Int] = None

  override def maximum: Option[Int] = None

  override def toString: String = "∅"

  override def equals(obj: Any): Boolean = obj match {
    case r: Range => r.isEmpty
    case _        => false
  }
}
